substrate/client/network/sync/src/strategy/state_sync.rs

```diff
@@ -132,6 +132,78 @@ where
 			skip_proof,
 		}
 	}
+
+	fn process_state_key_values(
+		&mut self,
+		state_root: Vec<u8>,
+		key_values: impl IntoIterator<Item = (Vec<u8>, Vec<u8>)>,
+	) {
+		let is_top = state_root.is_empty();
+
+		let (child_key_values, top_key_values): (Vec<_>, Vec<_>) =
+			key_values.into_iter().partition(|key_value| {
+				is_top && well_known_keys::is_child_storage_key(key_value.0.as_slice())
+			});
+
+		let entry = self.state.entry(state_root).or_default();
+
+		if entry.0.len() > 0 && entry.1.len() > 1 {
+			// Already imported child_trie with same root.
+			// Warning this will not work with parallel download.
+		} else {
+			self.imported_bytes +=
+				top_key_values.iter().map(|(key, _value)| key.len()).sum::<usize>() as u64;
+
+			entry.0.extend(top_key_values);
```

Comment:
We should not do the `partition` above and instead pass here the `iterator`. This way we would not need an extra allocation. Also we could do here a `filter_map` to filter out the child storage root.

Reply:
This piece has been rewritten in a plain for-loop, giving it a structure similar to the original too.